### ISSUE TITLE
Error: Az CLI Login failed. Please check the credentials and make sure az is installed on the runner.

### ISSUE THREAD

**Issue by mikhail-teslya** [Azure/login#205]
Hi, I also commented on a closed issue https://github.com/Azure/login/issues/203. This issue is not fixed for me in `@1.4.3` (and not working with `@v1` reference either). Error is exactly the same as in the first message.

1. Credentials are for sure working fine, executed `az login --service-principal -u <xx> --tenant <xx> -p <xx>` postfixed with `az account set --subscription <xx>` locally using the same creds which is in my repo secrets, cli login performed successfully
2. Attaching debug logs of workflow execution here (the main point was to find out az cli version for local comparison)
[azure-cli-workflow-run.txt](https://github.com/Azure/login/files/8132787/azure-cli-workflow-run.txt)

Please advise! Thanks

**Comment by BALAGA-GAYATRI** [comment]
Can you share your workflow yaml @mikhail-teslya ? That would help us know the issue better. Thanks!

**Comment by mikhail-teslya** [comment]
Hi @BALAGA-GAYATRI! Please find the workflow attached
[partial-web-app-remove-restrictions.txt](https://github.com/Azure/login/files/8175689/partial-web-app-remove-restrictions.txt)

**Comment by BALAGA-GAYATRI** [comment]
@mikhail-teslya Can you please confirm if this how you are generating credentials and putting them in secrets?
https://github.com/Azure/login#configure-a-service-principal-with-a-secret
If not please do this way.

**Comment by mikhail-teslya** [comment]
@BALAGA-GAYATRI yes, exactly the same way, as described on the page you've posted. I tried to regenerate credentials several times. And furthermore, I can login with the returned principal locally, as it is described in the first message in this thread

**Comment by thxmxx** [comment]
I have the same problem I just ran it from my repo and the exception is: `Error: Not all values are present in the credentials. Ensure clientId, clientSecret and tenantId are supplied.` 

But the JSON output from `az ad sp create-for-rbac` has the following keys:

```
{
  "appId": ...
  "displayName": ...,
  "name": ...,
  "password": ...,
  "tenant": ...
}
```

**Comment by BALAGA-GAYATRI** [comment]
@thxmxx  please provide ` --sdk-auth` in the end of your `rbac` command while generating the credentials as shown [here](https://github.com/Azure/login#configure-a-service-principal-with-a-secret). Currently, the object's format generated without this flag is not supported in the login action. Use `--sdk-auth` and it should solve the problem for you.

**Comment by thxmxx** [comment]
Yes, everything is working here, just telling to the author maybe he did the same mistake.

Ty

**Comment by mikhail-teslya** [comment]
Hello @BALAGA-GAYATRI, @thxmxx !
I did this many times before I created this issue, but just in case I repeated credentials creation step. Just in case, attaching a screenshot of the command I execute here:

![azure-credentials-creation](https://user-images.githubusercontent.com/85156799/156984177-5a5a049b-df2c-452f-b16d-b69cbde6b9e4.jpg)

Down below, you can see what it generates for me: 
```
{
  "clientId": "<my-client-id>",
  "clientSecret": "<the-client-secret>",
  "subscriptionId": "<my-correct-subscription-id>",
  "tenantId": "<some-guid-tenant-id>",
  "activeDirectoryEndpointUrl": "https://login.microsoftonline.com",
  "resourceManagerEndpointUrl": "https://management.azure.com/",
  "activeDirectoryGraphResourceId": "https://graph.windows.net/",
  "sqlManagementEndpointUrl": "https://management.core.windows.net:8443/",
  "galleryEndpointUrl": "https://gallery.azure.com/",
  "managementEndpointUrl": "https://management.core.windows.net/"
}
```
And after re-run, it fails with exactly the same error (please see logs in the first message). 
P.S. And yes, I did not forget to replace credentials in the github repo secrets (and it was the right secret) ;)

**Comment by mikhail-teslya** [comment]
@BALAGA-GAYATRI maybe it has something to do with https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-secrets, point **Never use structured data as a secret**?

**Comment by drr1217** [comment]
@BALAGA-GAYATRI I am receiving the same error. It indicates that the "--sdk-auth" has been deprecated and my output matches exactly what is seen from @mikhail-teslya.

**Comment by mikhail-teslya** [comment]
..and my secret name is AZURE_CREDENTIALS
![image](https://user-images.githubusercontent.com/85156799/157823199-69f72953-b9a1-4389-931b-7d2347e1c901.png)

**Comment by drr1217** [comment]
Just an update: changing the "uses" to "azure/actions/login@v1" seems to work, but I'm not sure why. This still seems like a bug to me.

<img width="238" alt="image" src="https://user-images.githubusercontent.com/19435583/157974893-6590f0c2-c1aa-4e24-9345-9ef82880f879.png">

**Comment by mikhail-teslya** [comment]
@BALAGA-GAYATRI will there be an update on that? It's hanging in `waiting-for-customer` for more than a week now..

**Comment by BALAGA-GAYATRI** [comment]
@drr1217  v1 always points to the latest release (v1.4.3). There shouldn't be any issue with that.

**Comment by BALAGA-GAYATRI** [comment]
@mikhail-teslya We are working on the issue and keep you posted.

**Comment by drr1217** [comment]
@BALAGA-GAYATRI Agreed, however, before I had "azure/login@v1" and received the error. When I changed it to "azure/actions/login@v1" the error no longer occurs. I was never pinning the version 1.4.3. 
@mikhail-teslya - just curious, does changing the action to "azure/actions/login@v1"fix the error for you too?

**Comment by mikhail-teslya** [comment]
Hi @drr1217 the answer is in the issue https://github.com/Azure/login/issues/203 ;)

**Comment by BALAGA-GAYATRI** [comment]
@mikhail-teslya looks like you are running on portal's az cli which might have older incompatible version. I would suggest you to do `az login` and create creds using `rbac` from your local machine and let us know the results once. Ideally this should work.

**Comment by mikhail-teslya** [comment]
Hi @BALAGA-GAYATRI !
Error is still the same, so it did not help.
Please check the below:
![azure-credentials-creation](https://user-images.githubusercontent.com/85156799/158785844-1461c880-8b0e-408a-b2d9-97e715a6a4ec.jpg)
credentials structured: 
```
{
  "clientId": "<my-client-id>",
  "clientSecret": "<my-client-secret>",
  "subscriptionId": "<the-corrct-subscription-id>",
  "tenantId": "<some-guid-tenant-id>",
  "activeDirectoryEndpointUrl": "https://login.microsoftonline.com",
  "resourceManagerEndpointUrl": "https://management.azure.com/",
  "activeDirectoryGraphResourceId": "https://graph.windows.net/",
  "sqlManagementEndpointUrl": "https://management.core.windows.net:8443/",
  "galleryEndpointUrl": "https://gallery.azure.com/",
  "managementEndpointUrl": "https://management.core.windows.net/"
}
```
Debug logs (login step):
[2_Run azurelogin@v1.4.3.txt](https://github.com/Azure/login/files/8283796/2_Run.azurelogin%40v1.4.3.txt)

The workflow itself is not changed

Please let me know if you need anything else

**Comment by BALAGA-GAYATRI** [comment]
@mikhail-teslya  Any luck after switching  to `ubuntu-latest` or `macos-latest` instead of windows?

**Comment by BALAGA-GAYATRI** [comment]
@mikhail-teslya Can you help me with one thing? Can you downgrade CLI version to maybe 2.33.1 and create credentials and let me know the result?

**Comment by mikhail-teslya** [comment]
> @mikhail-teslya Any luck after switching to `ubuntu-latest` or `macos-latest` instead of windows?
Hi, just tried and error is still the same

**Comment by BALAGA-GAYATRI** [comment]
Also try changing the resource group, subscription (create a new one may be) and change the SPN name also and see if it works. I just created new credentials and tested with both 2.34.1 and 2.33.1 CLI and both are working for me with `azure/login@v1` .

**Comment by mikhail-teslya** [comment]
Hi @BALAGA-GAYATRI! I can't change subscriptions easily, as it's managed by my organization. regarding different names - if you compare SPN names on screenshots [here](https://github.com/Azure/login/issues/205#issuecomment-1060260313) and [here](https://github.com/Azure/login/issues/205#issuecomment-1070758276), you'll see, that they're different. I can change it again, of course, but I can assure you I did this many times, before and after creating the issue, and it did not fix anything.

@tommasodotNET thank for the advice, I ran it and the output is:
![credentials-show](https://user-images.githubusercontent.com/85156799/158872017-f2012e2e-4a5d-4b6e-b5d9-5a23f0d8f846.jpg)

 So, I will try changing the resource group (and ok, as I still need new creds so I will re-create SPN with a different name one more time)

**Comment by mikhail-teslya** [comment]
@BALAGA-GAYATRI I found an issue, it's not directly related to the login action itself, but to how github manages secrets for callable workflows: secrets are not accessible inside a reusable workflow, please see https://github.com/actions/runner/issues/1490. After passing secrets directly to a callable workflow from the calling workflow, login was successful. Thanks to all who got a moment trying to help me out 👍

**Comment by captainjono** [comment]
@BALAGA-GAYATRI we are experiencing the SAME using a `2.33.1` using `ClientSecret` + `Service Principle`

`Azure/Login@V1` will NOT WORK ever!!!

`Azure/actions/login@V1` WORKS fine.

**Comment by AhirGhosh** [comment]
Also to add to the above comment, Azure/actions/login@V1 is working but generating an 'az upgrade' warning on ubuntu-latest.
So alongside the bug fix, upgrades to the GitHub hosted runners are also desirable for azure cli.

![image](https://user-images.githubusercontent.com/40377866/159601325-7b06a597-a26b-4a5f-b9a1-37c4436b6921.png)

**Comment by BALAGA-GAYATRI** [comment]
@mikhail-teslya You can use the secrets in the local actions/reusable workflows as well but you need to pass them as environment variables. Anyways good to know that your issue was resolved. :)

**Comment by roalexan** [comment]
FWIW, I hit the same problem and struggled through it, following various bits and pieces of what I read in this bug and [this one](https://github.com/Azure/login/issues/203) as well. The root cause for me was that the JSON that you paste into the GitHub "repository secret" MUST have the keys exactly as shown:

{
	"clientId": "AAA",
	"clientSecret": "BBB",
	"tenantId": "CCC",
	"subscriptionId": "DDD"
}

not sure if the order matters (probably not), but originally the output of the service principal that I created had a different set of keys - e.g.:

{
	"appId": "AAA",
	"name": "BBB",
	"password": "CCC",
	"tenant": "DDD"
}

so, I "transformed" the JSON - removed "name", added "subscriptionId", and changed the names of the other three keys  - and it is working for me now.

**Comment by BALAGA-GAYATRI** [comment]
@captainjono @AhirGhosh 
**There is no need to use `azure/actions/login@v1` as `azure/login@v1` is working perfectly fine for both OIDC and non-OIDC mechanisms**. This issue was initially labeled as `bug` because the reason was not known why the issue's author is not able to login. Now we got to know that it was [because of reusable workflows](https://github.com/Azure/login/issues/205#issuecomment-1075293393) which is not again related to `azure/login` action. Here are a few resources that might help you to use it.
1. [Sample workflow for using non-OIDC orSPN login](https://github.com/Azure/login#sample-workflow-that-uses-azure-login-action-to-run-az-cli)
2. [Sample workflow for using OIDC login](https://github.com/Azure/login#sample-workflow-that-uses-azure-login-action-using-oidc-to-run-az-cli-linux)
3. [Configure deployment credentials for both OIDC and non-OIDC](https://github.com/Azure/login#configure-deployment-credentials)

If you still face the issue, please share your workflows and error screenshots/logs for this action. We will be happy to help you. Otherwise, there is no such bug in this action.

**Comment by roalexan** [comment]
@BALAGA-GAYATRI thanks. yes, creating a new service principal via "az ad sp create-for-rbac" with the --sdk-auth option will output the correct keys. In my case, I was using an existing service principal and did not want to generate a new one, so just pointing out how I got that to work for me.

**Comment by alexanderdavide** [comment]
For me, the problem was that I was testing locally with [act](https://github.com/nektos/act) which doesn't have the Azure CLI installed as opposed to [Ubuntu based GitHub Shared Runners](https://github.com/actions/runner-images/blob/bbc911bacaf802fe776bc49536afa481adf6d0f6/images/linux/Ubuntu2004-Readme.md#cli-tools). In such cases, simply [install Azure CLI](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli-linux?pivots=apt#option-1-install-with-one-command) beforehand.

**Comment by robguy21** [comment]
In case someone else stumbles across this: my problem was that someone sent me a json payload from powershell with the `“` instead of `"` so my json wasn't valid which failed the authentication.

**Comment by dotjohann** [comment]
Hey,

I've followed this feed, still getting this error : `Error: Az CLI Login failed. Please check the credentials and make sure az is installed on the runner.`, tried everything and still fails, here is my setup :

>     - name: Azure Login
>       uses: azure/login@v1.4.6
>       with:
>         creds: ${{ secrets.AZURE_CREDENTIALS2 }}

Version on self-hosted runner : `"azure-cli": "2.40.0"`. Created the SPn - 3 times but the same error, with this command : `az ad sp create-for-rbac --name "ABCXYZ" --role contributor --scopes /subscriptions/ABC123/resourceGroups/XYZ123 --sdk-auth`.

Any ideas?

**Comment by bsm-github-jp** [comment]
For others having similar issues our use case is Azure Kubernetes Service deploying via github actions using OIDC. When you use the wizard to add a new workflow it creates a github secret named AZURE_CLIENT_ID each time you create a new workflow. Great if you only have one of them, in our use case we have several and had to create a custom secret for AZURE_CLIENT_ID such as AZURE_CLIENT_ID_PROD for each worfkflow. The value for the secret can be found as he application ID (Client ID) in Azure then app registrations.  Review the apps defined to locate you relevant client Id value.  Update your wokflow files in github to reference the updated secret name AZURE_CLIENT_ID_PROD and in our case things worked as expected. Hope this helps others out there having challenges, struggled with this one for a bit and it was less than fun. 

This can also be done with a single app registration, but that is outside the scope a bit, merely FYI.

**Comment by vancejc** [comment]
> For me, the problem was that I was testing locally with [act](https://github.com/nektos/act) which doesn't have the Azure CLI installed as opposed to [Ubuntu based GitHub Shared Runners](https://github.com/actions/runner-images/blob/bbc911bacaf802fe776bc49536afa481adf6d0f6/images/linux/Ubuntu2004-Readme.md#cli-tools). In such cases, simply [install Azure CLI](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli-linux?pivots=apt#option-1-install-with-one-command) beforehand.

For those who may have stumbled across this like I did, alternatively you can utilize the `-P ubuntu-latest=nektos/act-environments-ubuntu:18.04` flag for `act` which utilizes a 'full' ubuntu runner - https://github.com/nektos/act#default-runners-are-intentionally-incomplete

**Comment by sandervaneck** [comment]
Followed all the steps above, and it has always worked before. 
First I was getting: "no subscription I'd found"
Then I added the "no-subscriptions-allowed: true"
However, I'm now getting a new error in the authentication process of azure:
Invalid client secret provided. 

I am pasting the exact JSON that I've always done, with client Id, client secret, tenant id etc 

@BALAGA-GAYATRI could you help?

**Comment by ldvy** [comment]
As this seems to be the original issue created on this topic, I will also leave my comment here.
I have used the azure/login@v1 action for quite some time, and it always worked fine. However, today it did not...

I created my service principal the same way I did before (triple-checked it and recreated the SP two times to be sure) and created a repository secret with the JSON I got back from az cli, but it didn't work.
The JSON is valid and has all the necessary information.

To further prove this point, I am using the same SP to apply my Terraform config via GitHub Actions and hashicorp/setup-terraform@v2 runner - it works perfectly, so SP is not the issue.

Not sure why this issue was marked as Closed and "not a bug", when clearly it hasn't been fixed.

**Comment by YanaXu** [comment]
Hi @sandervaneck & @Idvy,

Since it's a closed issue, could you open 2 new issues for your cases?
You can link this issue to your new issue if necessary. And please provide the details in your issue. 
Just FYI, a "closed" issue may be missed and not be handled.

**Comment by elverado** [comment]
Brand new setup, all commands run and verified.

I used --sdk-auth which has a warning that it is deprecated, but loaded that into my Github Actions secret. Step with azure/login@v1 is failing to log in on a local windows runner with latest Azure CLI install. Still trialing some things like removing 32 bit Azure CLI and installing 64 bit.

**Comment by KiranKumarGowda** [comment]
Not all parameters are provided in 'creds'. Double-check if all keys are defined in 'creds': 'clientId', 'clientSecret', 'tenantId'.. Make sure 'az' is installed on the runner. If 'enable-AzPSSession' is true, make sure 'pwsh' is installed on the runner together with Azure PowerShell module. Double check if the 'auth-type' is correct

I'm facing this issues while deploying standard logic app using Github actions..Cam anyone help me out this.

**Comment by KiranKumarGowda** [comment]
> Not all parameters are provided in 'creds'. Double-check if all keys are defined in 'creds': 'clientId', 'clientSecret', 'tenantId'.. Make sure 'az' is installed on the runner. If 'enable-AzPSSession' is true, make sure 'pwsh' is installed on the runner together with Azure PowerShell module. Double check if the 'auth-type' is correct
> 
> I'm facing this issues while deploying standard logic app using Github actions..Cam anyone help me out this.

@BALAGA-GAYATRI could you help me out please.

**Comment by YanaXu** [comment]
@KiranKumarGowda , please open a new issue to track your question. We may miss the comments from a closed issue.
And in your new issue, please provide your workflow fie and the debug log.

**Comment by KiranKumarGowda** [comment]
> @KiranKumarGowda , please open a new issue to track your question. We may miss the comments from a closed issue. And in your new issue, please provide your workflow fie and the debug log.

Ok give me a minute

**Comment by mumian** [comment]
I ran into the same error message - "Login failed with Error: Content is not a valid JSON object. Double check if the 'auth-type' is correct."  The problem was that I left a comma at the end of the tenantId line when I trimmed the "az ad sp create-for-rbac".  

{
  "clientId": "<my-client-id>",
  "clientSecret": "<the-client-secret>",
  "subscriptionId": "<my-correct-subscription-id>",
  "tenantId": "<some-guid-tenant-id>",
}

The error message should be improved.
